### ISSUE TITLE
Add random jitter before taking action

### ISSUE THREAD

**Issue by therc** [kelseyhightower/confd#400]
To help prevent thundering herd scenarios, allow waiting a random number between 0 and X (or between X and Y) of seconds before generating/reloading a file.

**Comment by bacongobbler** [comment]
@therc you mean `-i/--interval`?

**Comment by therc** [comment]
-i is the polling interval. You can use that to approximate random jitter for the single file case. But if you have multiple files getting generated with different jitter ranges or want to use a watch, -i is not going to work.

**Comment by bacongobbler** [comment]
> But if you have multiple files getting generated with different jitter ranges or want to use a watch

Confd doesn't work on config-per-file. In that case we have suggested previously that if people want different jitter ranges for different files that they should use a separate confd process with different config set up. That would be the simplest solution out of band without convoluting the codebase.

**Comment by bacongobbler** [comment]
Ah I see what you mean. You mean by the fact that each template resource being processed = a call to the backend, meaning that number of connections to the backend increases based on the number of templates you have. We could probably optimize this instead by grabbing all gets from the store at once and injecting them into the templates at once before processing, preventing the thundering herd problem.

Would that work for you?

**Comment by therc** [comment]
That's one of the scenarios, yes, but the first I thought of was:

Put a watch on K
Get woken up, process template
Restart/Signal daemon D
D acts on the new configuration, which makes it fetch something from another service, S
S gets hammered by a hundred instances of D
???
No profit!

Actually, even with -i, if we assume that on something like Kubernetes all pods will initially get scheduled with good probability during the same few seconds, it might make sense to wait for a random interval before polling for the first time.

**Comment by therc** [comment]
Multiple confd instances is not ideal, but I guess I could live with it in some cases. The fact that they're not coordinated is an asset for multiple jitter ranges, but it's also a liability when it comes to predicting behaviour, especially the worst case. Oh well. :-)

**Comment by JacobHayes** [comment]
This would be very helpful for a group of nodes that all watch the same keys and are restarted when the template updates. For example, a cluster of web servers using config files generated by confd. All the instances of confd watch the same keys, so when it's updated all servers are restarted around the same time. A jitter or delay option in the template resource would be helpful to spread the downtime across these nodes. It'd be interesting if it chose a delay within an interval (ex: `[1,5]`) between that before triggering the `reload_cmd`.

I guess an alternative would be to implement that functionality in the `reload_cmd` itself.

**Comment by bacongobbler** [comment]
I think it'd be better to implement that in the reload command. Users want immediate updates when keys in etcd change, such as an `nginx reload`. Technically if your process listens to SIGHUP and does a proper reload rather than a restart then this wouldn't be an issue.

**Comment by JacobHayes** [comment]
Yeah, immediate updates would still happen without that set. That's true, but in some cases not all changes can made with just a reload. Anyway, I agree that it can be pushed down to `reload_cmd`.
